go-controller/pkg/ovn/controller/network_qos/network_qos_controller.go

```diff
@@ -0,0 +1,452 @@
+	defer utilruntime.HandleCrash()
+
+	klog.Infof("Starting controller %s", c.controllerName)
+
+	// Wait for the caches to be synced
+	klog.V(5).Info("Waiting for informer caches to sync")
+	if !util.WaitForInformerCacheSyncWithTimeout(c.controllerName, stopCh, c.nqosCacheSynced, c.nqosNamespaceSynced, c.nqosPodSynced, c.nqosNodeSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		klog.Errorf("Error syncing caches for network qos")
+		return
+	}
+
+	klog.Infof("Repairing Network QoSes")
+	// Run the repair function at startup so that we synchronize KAPI and OVNDBs
+	err := c.repairNetworkQoSes()
+	if err != nil {
+		klog.Errorf("Failed to repair Network QoS: %v", err)
+	}
+
+	wg := &sync.WaitGroup{}
+	// Start the workers after the repair loop to avoid races
+	klog.V(5).Info("Starting Network QoS workers")
+	for i := 0; i < threadiness; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			wait.Until(func() {
+				c.runNQOSWorker(wg)
+			}, time.Second, stopCh)
+		}()
+	}
+
+	klog.V(5).Info("Starting Namespace Network QoS workers")
+	for i := 0; i < threadiness; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			wait.Until(func() {
+				c.runNQOSNamespaceWorker(wg)
+			}, time.Second, stopCh)
+		}()
+	}
+
+	klog.V(5).Info("Starting Pod Network QoS workers")
+	for i := 0; i < threadiness; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			wait.Until(func() {
+				c.runNQOSPodWorker(wg)
+			}, time.Second, stopCh)
+		}()
+	}
+
+	klog.V(5).Info("Starting Node Network QoS workers")
+	for i := 0; i < threadiness; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			wait.Until(func() {
+				c.runNQOSNodeWorker(wg)
+			}, time.Second, stopCh)
+		}()
+	}
+
+	<-stopCh
+
+	klog.Infof("Shutting down controller %s", c.controllerName)
+	c.nqosQueue.ShutDown()
+	c.nqosNamespaceQueue.ShutDown()
+	c.nqosPodQueue.ShutDown()
+	c.teardownMetricsCollector()
+	wg.Wait()
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and
+// marks them done. You may run as many of these in parallel as you wish; the
+// workqueue guarantees that they will not end up processing the same object
+// at the same time.
+func (c *Controller) runNQOSWorker(wg *sync.WaitGroup) {
+	for c.processNextNQOSWorkItem(wg) {
+	}
+}
+
+func (c *Controller) runNQOSNamespaceWorker(wg *sync.WaitGroup) {
+	for c.processNextNQOSNamespaceWorkItem(wg) {
+	}
+}
+
+func (c *Controller) runNQOSPodWorker(wg *sync.WaitGroup) {
+	for c.processNextNQOSPodWorkItem(wg) {
+	}
+}
+
+func (c *Controller) runNQOSNodeWorker(wg *sync.WaitGroup) {
+	for c.processNextNQOSNodeWorkItem(wg) {
+	}
+}
+
+// handlers
+
+// onNQOSAdd queues the NQOS for processing.
+func (c *Controller) onNQOSAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+	klog.V(4).Infof("Adding Network QoS %s", key)
```

Comment:
as we scale up number of controllers (UDNs/NADs) wont this print that many times?

Reply:
removed this.